ADR/00xx-related-pipelines.md

```diff
@@ -0,0 +1,103 @@
+  * This is a one-to-many dependency. One operator bundle depends on many operand images.
+* A user team just has **two components that depend on each other in a functional way**. When they propose an update to one component, they want to submit a corresponding change to the second component, and have those tested together before merging both.
+  * In this case, on any given day, component B’s PRs may or may not have dependencies on PRs from component A. Not all changes are linked.
+  * This is a many-to-many dependency. The user may have lots of components that depend on lots of other components.
+
+In theory **any combination of the cases above** could be present in an app. They could have one common parent image, many operand images that depend on that, and one or more operator bundles which depend on those.
+
+Today, users work around how complicated it is to manage digests themselves by instead using *floating tags*, which have the benefit of being easy to use - no need to update - but have the problem of being unclear. It's not exactly clear what you're building against if you refer to it by tag. Potentially insecure. We want to make it easy for users to do better.
+
+## Decision
+
+In order to support linking of multiple components, any component dependency will be encoded as part of the Tekton definition to be run for **pull request** event types.
+
+### User flow
+
+Users will be able to specify dependencies between components by modifying the pull request Tekton definition or by adding configuration to the UI which will result in AppStudio encoding the configuration in the Tekton definition. This definition will allow the specification of whether dependent components should be run in parallel or serially when combined.
+
+### Pipelines as Code (PaC)
+
+Pipelines as Code (PaC) will expand its design to include orchestration logic. This orchestrator will enable a plugable scheduler which is capable of identifying the dependent repositories and Tekton definitions (composing the metadata required to specify a component in AppStudio). When a component's Tekton definition indicates a dependency on another component, PaC will be able to combine the defined pipelines so that the resulting Pipeline definition builds images in the proper order.
+
+When identifying the pipeline definitions to merge, PaC will first look at the linked components for the same branch where a build was triggered from. If no matching branches are found then the default branch will be used.
+
+PaC will need to recursively expand the directed acyclic graph (DAG) to ensure that all dependent components are linked.
+
+### Build service
+
+The build service will be able to detect any component pull specs from earlier in a composite pipeline that differs from a previous Snapshot and replace all references of the previous pull spec with the new pull spec before building the component.
```

Comment:
In https://github.com/redhat-appstudio/book/pull/114#discussion_r1254871127, @ralphbean commented:

> I'm not sure how build-service will do this.
> 
> The work to be done: "replace some pullspec references with others"
> 
> * Should it be the responsibility of tasks in the composite pipeline?
> * Should it be the responsibility of PaC when assembling the composite pipeline?
> 
> I don't think build-service can intervene at composite-pipeline-assembly time or at composite-pipeline-run-time.

Reply:
I was thinking that the functionality would be somewhat similar to how Cachi2 has been implemented where the Dockerfile is modified based on the prefetched input: https://github.com/redhat-appstudio/build-definitions/blob/main/task/buildah/0.1/buildah.yaml#L146-L152

The build service would have to ingest the results of the previous tasks and replace the references matching those earlier components. For common parent images, I would expect the images from the workspace's quay organization to be leveraged so it should be easy to detect required updates. For bundle images where the image references may be present with modified pull specs (i.e. preparing for a release), the replacement may be harder unless the build service has access to the latest Snapshot or Global Candidate List so that it can determine what the _previous/latest_ digest of the rebuilt image is and then substitute solely based on digest matching.